sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DataStreams.java

```diff
@@ -186,14 +186,16 @@ public DataStreamDecoder(Coder<T> coder, PrefetchableIterator<ByteString> inputS
      * ByteString} in the underlying {@link ByteString} {@link Iterator iterator} and decoding
      * elements till at the next boundary.
      */
-    public List<T> decodeFromChunkBoundaryToChunkBoundary() {
+    public WeightedList<T> decodeFromChunkBoundaryToChunkBoundary() {
       inbound.currentStream = inputByteStrings.next().newInput();
       inbound.position = 0;
       try {
         InputStream previousStream = inbound.currentStream;
-        List<T> rvals = new ArrayList<>();
+        WeightedList<T> rvals = new WeightedList<>(new ArrayList<>(), 0L);
         while (previousStream == inbound.currentStream && inbound.currentStream.available() != 0) {
-          rvals.add(next());
+          inbound.resetCounter();
+          T next = next();
+          rvals.add(next, inbound.getCount());
```

Comment:
We could add 8 to the weight (to account for the contribution of placing it in the list). This would at least help the varint case be only a 2-3x underestimate, rather than a 16-24x one.